### PR TITLE
billing: usage notifications

### PR DESCRIPTION
**Pull request by gioelecerati**
**What does this pull request do? Explain your changes. (required)**

**Specific updates (required)**

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Screenshots (optional)**

**Checklist**

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/api/src/controllers/notification.ts
Comment (victorges) [inline packages/api/src/controllers/notification.ts:1]: WDYT of moving this file to a folder like `./controllers/utils` or move the helpers to `helpers.ts`? The `controllers` folder is meant to hold only files matching APIs, with the only exception today being that existing `helpers.ts` file. 

In the future would make sense to move the generic `helpers` to to separate folder as well, tho I think it's enough for now just moving this to that separate folder.

File: packages/api/src/schema/db-schema.yaml [hunk 1]
```diff
@@ -1060,9 +1060,22 @@ components:
         migrationInvoice:
           type: boolean
           default: false
-        lastEmailAboutPaymentFailure:
-          type: number
-          example: 1587667174725
+        notifications:
```
Comment (victorges) [inline packages/api/src/schema/db-schema.yaml:1075]: neat

File: packages/api/src/controllers/usage.ts [hunk 2]
```diff
@@ -2,10 +2,12 @@ import Router from "express/lib/router";
 import { db } from "../store";
 import { authorizer, validatePost } from "../middleware";
 import { products } from "../config";
-import { reportUsage } from "./stripe";
 import fetch from "node-fetch";
 import qs from "qs";
 import { NotFoundError } from "../store/errors";
+import { WithID } from "../store/types";
+import { User } from "../schema/types";
```
Comment (victorges) [inline packages/api/src/controllers/usage.ts:9]:
```suggestion
import { User, WithID } from "../store/types";
```

File: packages/api/src/controllers/usage.ts [hunk 3]
```diff
@@ -63,6 +65,68 @@ export const calculateOverUsage = async (product, usage) => {
...
+    TotalUsageMins: Math.max(
+      (usage?.TotalUsageMins / (limits.transcoding || 0)) * 100,
+      0
+    ),
+    DeliveryUsageMins: Math.max(
+      (usage?.DeliveryUsageMins / (limits.streaming || 0)) * 100,
+      0
+    ),
+    StorageUsageMins: Math.max(
+      (usage?.StorageUsageMins / (limits.storage || 0)) * 100,
```
Comment (victorges) [inline packages/api/src/controllers/usage.ts]: Risk of dividing by `0` in all these. perhaps `1`? Or maybe you want a top-level ternary on `limits.transcoding ? ...` instead

File: packages/api/src/controllers/usage.ts [hunk 4]
```diff
@@ -63,6 +65,68 @@ export const calculateOverUsage = async (product, usage) => {
   return overUsage;
 };
 
+export const getUsagePercentageOfLimit = async (product, usage) => {
+  let limits: any = {};
```
Comment (victorges) [inline packages/api/src/controllers/usage.ts]: ts to the rescue (from `any`) 😆 
```suggestion
  let limits: Record<string, number> = {};
```

File: packages/api/src/controllers/notification.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
+import { sendgridEmail } from "./helpers";
+import { frontendUrl, unsubscribeUrl } from "./user";
+import { db } from "../store";
+import { sendgridEmailPaymentFailed } from "./helpers";
+import { WithID } from "../store/types";
+import { Request } from "express";
+import { User } from "../schema/types";
+
+export const getUsageNotifications = async (
+  usagePercentages,
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: Can you add a type to this? Can even inline it e.g.
```suggestion
  usagePercentages: {TotalUsageMins: number, DeliveryUsageMins: number, StorageUsageMins: number},
```

File: packages/api/src/controllers/notification.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  user: WithID<User>
+) => {
+  let notifications = [];
+
+  if (
+    usagePercentages.TotalUsageMins > 90 ||
+    usagePercentages.DeliveryUsageMins > 90 ||
+    usagePercentages.StorageUsageMins > 90
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: I think these fields have confusing names. They imply the unit is minutes, when I believe it is actually a percentage? WDYT of renaming to e.g. `TotalUsageLimitPercent`, `just `...Percent` or sth like that?

File: packages/api/src/controllers/notification.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  ) {
+    notifications.push({
+      type: "notification90",
+      title: "Usage Warning",
+      message: `Your usage is over 90% of your limit.`,
+    });
+  } else if (
+    usagePercentages.TotalUsageMins > 75 ||
+    usagePercentages.DeliveryUsageMins > 75 ||
+    usagePercentages.StorageUsageMins > 75
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: An alternative is grabbing the `Math.max` first then checking once each time

File: packages/api/src/controllers/notification.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  if (
+    usagePercentages.TotalUsageMins > 90 ||
+    usagePercentages.DeliveryUsageMins > 90 ||
+    usagePercentages.StorageUsageMins > 90
+  ) {
+    notifications.push({
+      type: "notification90",
+      title: "Usage Warning",
+      message: `Your usage is over 90% of your limit.`,
+    });
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: shouldnt you update the user object flag here?

File: packages/api/src/controllers/notification.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  } else if (
+    usagePercentages.TotalUsageMins > 75 ||
+    usagePercentages.DeliveryUsageMins > 75 ||
+    usagePercentages.StorageUsageMins > 75
+  ) {
+    await db.user.update(user.id, {
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: shouldnt you push a notification here? 

also, when are the `notifications.usage.notification75/90` checked?

File: packages/api/src/controllers/notification.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    if (
+      user.notifications?.usage?.notification75 ||
+      user.notifications?.usage?.notification90
+    ) {
+      await db.user.update(user.id, {
+        notifications: {
+          usage: {
+            notification75: false,
+            notification90: false,
+          },
```
Comment (victorges) [inline packages/api/src/controllers/notification.ts]: This seems unnecessary, could add a comment explaining why do you need this?

**Review by victorges — APPROVED** [review]
LGTM!

File: packages/api/src/controllers/usage.ts (hunk 3 above)
Comment (gioelecerati) [reply packages/api/src/controllers/usage.ts]: ouch! fixed

File: packages/api/src/controllers/notification.ts (hunk 9 above)
Comment (gioelecerati) [reply packages/api/src/controllers/notification.ts]: Oh right, it was mixed up, all db.set and all checks are done in the function `notifyUser`

```
export const notifyUser = async (notifications, user, req) => {
```

File: packages/api/src/controllers/notification.ts (hunk 8 above)
Comment (gioelecerati) [reply packages/api/src/controllers/notification.ts]: it's happening in `notifyUser`

File: packages/api/src/controllers/notification.ts (hunk 10 above)
Comment (gioelecerati) [reply packages/api/src/controllers/notification.ts]: This is resetting the notifications sent to false for each type when all three types of usage are under 75% (so at the reset of the billing cycle)

**Pull request merged**